File: Src/Menu.hs
module Src.Menu where

    import Src.Controller.AlunoController
    import Src.Controller.ChatController
    import Src.Controller.MonitorController
    import Src.Controller.ProfessorController
    import Src.Util.TxtFunctions
    import Src.Model.Monitor as M
    import Src.Model.Professor as P
    import Src.Model.Aluno as A

    menuPrincipal :: IO()
    menuPrincipal = do
        putStrLn "\nBem vindo ao SAD: Sistema de Atendimento ao Discente! :):"
        menuLogin
    
    decideMenuPrincipal :: String -> IO ()
    decideMenuPrincipal opcao
        | opcao == "1" = menuLogin
        | opcao == "2" = putStrLn "Saindo..."
        | otherwise = do 
            putStrLn "Insira um valor válido!\n"
            menuPrincipal

    menuCadastro :: Professor -> IO ()
    menuCadastro professor = do
        putStrLn "\nQuem você deseja vincular?"
        putStrLn "1) Vincular aluno\n2) Vincular monitor\n3) Voltar para o menu professor"
        opcao <- getLine
        putStr "\n"
        decideMenuCadastro professor opcao
    
    decideMenuCadastro :: Professor -> String -> IO ()
    decideMenuCadastro professor opcao 
        | opcao == "1" = do
            disciplina <- solicitaDisciplina professor
            adicionaAluno disciplina
        | opcao == "2" = do
            disciplina <- solicitaDisciplina professor
            adicionaMonitor disciplina
        | opcao == "3" = putStrLn ""
        | otherwise = do
            putStrLn "Insira um valor válido!\n"
            menuCadastro professor

    menuLogin :: IO()
    menuLogin = do
        putStrLn "Insira seu ID para entrar. Para sair do sistema, digite SAIR:"

        input <- getLine
        putStr "\n"

        if input == "SAIR" then putStrLn "Saindo..." else do
            let idPerfil = read input :: Int

            ehMonitor <- checaExistenciaById "Monitores" idPerfil
            ehProfessor <- checaExistenciaById "Professores" idPerfil
            ehAluno <- checaExistenciaById "Alunos" idPerfil

            decideMenuLogin idPerfil ehMonitor ehProfessor ehAluno

    decideMenuLogin :: Int -> Bool -> Bool -> Bool -> IO()
    decideMenuLogin idPerfil ehMonitor ehProfessor ehAluno
        | ehMonitor = decideMenuAlunoMonitor idPerfil
        | ehAluno = exibeMenuAluno idPerfil
        | ehProfessor = exibeMenuProfessor idPerfil
        | otherwise = do
            putStrLn "Matrícula/ID não encontrado. Tente novamente\n\n"
            menuLogin

    decideMenuAlunoMonitor :: Int -> IO()
    decideMenuAlunoMonitor idPerfil = do
        monitor <- getMonitor idPerfil
        putStrLn ("Foi identificado que você é monitor da disciplina: " ++ disciplina monitor)
        putStrLn ("Como deseja entrar no sistema?\n\n1) Entrar como Aluno\n2) Entrar como Monitor de " ++ disciplina monitor)
        escolha <- getLine
        putStr "\n"
        if escolha == "1" then exibeMenuAluno idPerfil else 
            if escolha == "2" then exibeMenuMonitor idPerfil else do
                putStrLn "Insira um valor válido!"
                decideMenuAlunoMonitor idPerfil

    exibeMenuProfessor :: Int -> IO()
    exibeMenuProfessor idProfessor = do
        professor <- getProfessor idProfessor
        putStrLn "\n== SAD: MENU PROFESSOR =="
        putStrLn ("ID: " ++ show (P.id professor) ++ " | " ++ "Nome: " ++ P.nome professor ++ " | " ++ "Disciplinas: " ++ show (P.disciplinas professor))
        putStrLn "Digite o número da ação que deseja executar!\n"
        putStrLn "1) Exibir tickets\n2) Responder Tickets em progresso\n3) Vincular aluno/monitor\n4) Desvincular aluno/monitor\n5) Deslogar"
        opcao <- getLine
        decideMenuProfessor professor opcao

    decideMenuProfessor :: Professor -> String -> IO()
    decideMenuProfessor professor opcao
        | opcao == "1" = do
            lerTicketsDisciplinaProfessor professor
            exibeMenuProfessor (P.id professor)
        | opcao == "2" = do
            putStrLn "Respondendo tickets...\n"
            exibeMenuProfessor (P.id professor)
        | opcao == "3" = do
            menuCadastro professor
            exibeMenuProfessor (P.id professor)
        | opcao == "4" = do
            removeMonitor (P.disciplinas professor)
            exibeMenuProfessor (P.id professor)
        | opcao == "5" = do
            putStrLn "Deslogando...\n"
            menuPrincipal
        | otherwise = do 
            putStrLn "Opção inválida!"
            exibeMenuProfessor (P.id professor)

    exibeMenuMonitor :: Int -> IO()
    exibeMenuMonitor id = do
        monitor <- getMonitor id
        aluno <- getAluno id
        putStrLn "\n== SAD: MENU MONITOR =="
        putStrLn ("ID: " ++ show (M.id monitor) ++ " | " ++ "Nome: " ++ A.nome aluno ++ " | " ++ "Disciplina: " ++ M.disciplina monitor)
        putStrLn "Digite o número da ação que deseja executar!\n"
        putStrLn "1) Exibir todos os tickets\n2) Responder tickets em andamento\n3) Deslogar"
        opcao <- getLine 
        decideMenuMonitor monitor opcao

    decideMenuMonitor :: Monitor -> String -> IO()
    decideMenuMonitor monitor opcao
        | opcao == "1" = do
            exibeMensagensDisciplina (disciplina monitor)
            exibeMenuMonitor (M.id monitor)
        | opcao == "2" = do
            adicionaMensagemMonitor monitor
            exibeMenuMonitor (M.id monitor)
        | opcao == "3" = do
            putStrLn "Deslogando...\n"
            menuPrincipal
        | otherwise  = do
            putStrLn "Opção inválida!\n"
            exibeMenuMonitor (M.id monitor)

    exibeMenuAluno :: Int -> IO()
    exibeMenuAluno idAluno = do
        aluno <- getAluno idAluno
        putStrLn "\n== SAD: MENU ALUNO =="
        putStrLn ("ID: " ++ show (A.id aluno) ++ " | " ++ "Nome: " ++ A.nome aluno ++ " | " ++ "Disciplinas: " ++ show (A.disciplinas aluno))
        putStrLn "Digite o número da ação que deseja executar!\n"
        putStrLn "1) Matricular-se em disciplina\n2) Desmatricular-se de disciplina\n3) Ler tickets de uma disciplina\n4) Ler meus tickets\n5) Criar Ticket\n6) Mandar mensagem em um ticket meu\n7) Marcar ticket como resolvido\n8) Excluir ticket\n9) Deslogar"
        opcao <- getLine 
        decideMenuAluno aluno opcao
    
    decideMenuAluno :: Aluno -> String -> IO()
    decideMenuAluno aluno opcao
        | opcao == "1" = do
            matriculaAlunoEmDisciplina aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "2" = do
            desmatriculaAlunoDeDisciplina aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "3" = do
            leTicketsDaDisciplinaAluno aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "4" = do
            leTicketsDoAluno aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "5" = do
            adicionaTicket aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "6" = do
            adicionaMensagemAluno aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "7" = do
            resolveTicket aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "8" = do
            excluirTicket aluno
            exibeMenuAluno (A.id aluno)
        | opcao == "9" = do
            putStrLn "Deslogando...\n"
            menuPrincipal
        | otherwise  = do
            putStrLn "Opção inválida!\n"
            exibeMenuAluno (A.id aluno)

File: prolog/src/Menu.pl
:- module('Menu', [menuPrincipal/0]).
:- use_module('controller/MonitorController.pl', [vinculaMonitor/0, getMonitor/2, ehMonitor/1, desvinculaMonitor/0]).
:- use_module('controller/ChatController.pl', [exibeTicketsDisciplina/1, exibeTicketsAluno/1, responderTicket/2]).
:- use_module('controller/ProfessorController.pl', [getProfessor/2, ehProfessor/1]).
:- use_module('controller/AlunoController.pl', [getAluno/2, ehAluno/1]).
:- use_module('util/jsonFunctions', [checaExistencia/2, atualizaAtributoAluno/3, atualizaAtributoProfessor/3, getObjetoByID/3]).
:- use_module('util/EncriptFunctions.pl', [encripta/3]).

menuPrincipal() :- writeln('\n\nBem vindo ao SAD: Sistema de Atendimento ao Discente! :):'), menuLogin().

menuLogin() :- 
    writeln("Insira seu ID para entrar. Para sair do sistema, digite 'sair': "), 
    read(Id),
    decideMenu(Id).

menuAutenticacao(Objeto):-
    writeln("\nInsira a sua Senha:"),
    read(Senha),
    encripta(Senha, Objeto.nome, SenhaEncriptada),
    SenhaEncriptada = Objeto.senha.

decideMenu(sair):- halt(0).
decideMenu(Id) :- 
    atom_string(Id, IdString),
    (
      ehProfessor(IdString) -> 
        getProfessor(IdString, Professor), 
        (menuAutenticacao(Professor) -> exibeMenuProfessor(IdString); writeln("Senha incorreta\n"), menuLogin());
      ehMonitor(IdString) -> 
        getAluno(IdString, Monitor),
        (menuAutenticacao(Monitor) -> exibeMenuAlunoMonitor(IdString); writeln("Senha incorreta\n"), menuLogin());
      ehAluno(IdString) -> 
        getAluno(IdString, Aluno),
        (menuAutenticacao(Aluno) -> exibeMenuAluno(IdString); writeln("Senha incorreta\n"), menuLogin());
      writeln("ID nao encontrado!\n"),
      menuLogin()
    ).

trocarSenha(Entidade):-
    writeln('Digite sua nova senha:'),
    read(Senha),
    encripta(Senha, Entidade.nome, SenhaEncriptada),
    (
       ehAluno(Entidade.id) -> atualizaAtributoAluno(Entidade.id, "senha", SenhaEncriptada);
       atualizaAtributoProfessor(Entidade.id, "senha", SenhaEncriptada)
    ).

perguntaDisciplina(Disciplinas, Disciplina):-
    length(Disciplinas, Size), Size > 1,
    writeln("Insira a disciplina:"),
    read(SiglaDisciplina), nl,
    atom_string(SiglaDisciplina, SiglaString),
    (
        member(SiglaString, Disciplinas) -> Disciplina = SiglaString;
        Disciplina = "INVALIDA"
    )
    ;
    [H|_] = Disciplinas,
    Disciplina = H.

msgInputInvalido():- writeln("Insira um valor valido\n").

%----------------------------------------------------- PROFESSOR -----------------------------------------------------%

exibeMenuProfessor(Id):-
    getProfessor(Id, Professor),
    writeln('\n== SAD: MENU PROFESSOR =='),
    swritef(Out, "\nID: %w | Nome: %w | Disciplinas: %w\n", [Professor.id, Professor.nome, Professor.disciplinas]), write(Out),
    writeln('Digite o numero da acao que deseja executar!\n'),
    writeln('1) Exibir tickets\n2) Responder Tickets em andamento\n3) Vincular aluno/monitor\n4) Desvincular aluno/monitor\n5) Alterar senha de acesso\n6) Deslogar\n'),
    read(Opcao),
    decideMenuProfessor(Opcao, Professor),
    exibeMenuProfessor(Id).

decideMenuProfessor(1, Professor) :- 
    perguntaDisciplina(Professor.disciplinas, Disciplina), 
    (Disciplina = "INVALIDA" -> decideMenuProfessor(-1, Professor) ; exibeTicketsDisciplina(Disciplina)).
    

decideMenuProfessor(2, Professor):- 
    perguntaDisciplina(Professor.disciplinas, Disciplina), 
    (Disciplina = "INVALIDA" -> decideMenuProfessor(-1, Professor) ; responderTicket(Professor, Disciplina)).

decideMenuProfessor(3, Professor):- menuCadastroProfessor(Professor).

decideMenuProfessor(4, Professor):- menuRemocaoProfessor(Professor).

decideMenuProfessor(5, Professor):- 
    trocarSenhaProfessor(Professor),
    exibeMenuProfessor(Professor.id).

decideMenuProfessor(6, _) :- writeln('Deslogando...'), menuPrincipal().

decideMenuProfessor(_, _) :- msgInputInvalido().

menuCadastroProfessor(Professor) :- 
    writeln('\nQuem voce deseja vincular?'),
    writeln('1) Vincular aluno\n2) Vincular monitor\n3) Voltar para o menu professor\n'),
    read(Opcao),
    decideMenuCadastro(Opcao, Professor),
    exibeMenuProfessor(Professor.id).

decideMenuCadastro(1, _).

decideMenuCadastro(2, Professor):-
    vinculaMonitor(),
    exibeMenuProfessor(Professor.id).

decideMenuCadastro(3, _).

decideMenuCadastro(_, Professor):- msgInputInvalido(), menuCadastroProfessor(Professor).

menuRemocaoProfessor(Professor) :- 
    writeln('\nQuem voce deseja desvincular?'),
    writeln('1) Desvincular aluno\n2) Desvincular monitor\n3) Voltar para o menu professor\n'),
    read(Opcao),
    decideMenuRemocao(Opcao, Professor),
    exibeMenuProfessor(Professor.id).

decideMenuRemocao(1, _).

decideMenuRemocao(2, Professor) :-
    desvinculaMonitor(),
    exibeMenuProfessor(Professor.id).

decideMenuRemocao(3, Professor):- exibeMenuProfessor(Professor.id).

decideMenuRemocao(_, Professor):- msgInputInvalido(), menuRemocao(Professor).

%----------------------------------------------------- MONITOR -----------------------------------------------------%

exibeMenuAlunoMonitor(Id) :- 
    getMonitor(Id, Monitor),
    write('\nFoi identificado que voce eh monitor da disciplina: '), write(Monitor.disciplina),
    write('\nComo deseja entrar no sistema?\n\n1) Entrar como Aluno\n2) Entrar como Monitor de '), write(Monitor.disciplina), nl,
    read(Opcao),
    decideMenuAlunoMonitor(Opcao, Id).
  
decideMenuAlunoMonitor(1, Id) :- exibeMenuAluno(Id).
decideMenuAlunoMonitor(2, Id) :- exibeMenuMonitor(Id).
decideMenuAlunoMonitor(_, _) :- msgInputInvalido(), exibeMenuAlunoMonitor().

exibeMenuMonitor(Id) :- 
    getMonitor(Id, Monitor),
    getAluno(Id, Aluno),
    writeln('\n== SAD: MENU MONITOR =='),
    swritef(Out, '\nID: %w | Nome: %w | Disciplina: %w\n', [Monitor.id, Aluno.nome, Monitor.disciplina]), write(Out),
    writeln('Digite o numero da ação que deseja executar!\n'),
    writeln('1) Exibir todos os tickets\n2) Responder tickets em andamento\n3) Deslogar\n'),
    read(Opcao),
    decideMenuMonitor(Opcao, Monitor),
    exibeMenuMonitor(Id).

decideMenuMonitor(1, Monitor) :- exibeTicketsDisciplina(Monitor.disciplina).

decideMenuMonitor(2, Monitor):- responderTicket(Monitor, Monitor.disciplina).

decideMenuMonitor(3, _) :- write('\nDeslogando...'), menuPrincipal().

decideMenuMonitor(_) :- msgInputInvalido().

%----------------------------------------------------- ALUNO -----------------------------------------------------%

exibeMenuAluno(Id):-
    getAluno(Id, Aluno),
    writeln('\n== SAD: MENU ALUNO =='),
    swritef(Out, '\nID: %w | Nome: %w | Disciplinas: %w\n', [Aluno.id, Aluno.nome, Aluno.disciplinas]), write(Out),
    writeln('Digite o numero da acao que deseja executar!\n'),
    writeln('1) Ler tickets de uma disciplina\n2) Ler meus tickets\n3) Criar Ticket\n4) Mandar mensagem em um ticket meu\n5) Marcar ticket como resolvido\n6) Excluir ticket\n7) Trocar senha de acesso\n8) Deslogar\n'),
    read(Opcao),
    decideMenuAluno(Opcao, Aluno),
    exibeMenuAluno(Id).

decideMenuAluno(1, Aluno):- 
    perguntaDisciplina(Aluno.disciplinas, Disciplina), 
    (Disciplina = "INVALIDA" -> decideMenuAluno(-1, Aluno) ; exibeTicketsDisciplina(Disciplina)).

decideMenuAluno(2, Aluno):- exibeTicketsAluno(Aluno.id).

decideMenuAluno(3, _).

decideMenuAluno(4, _).

decideMenuAluno(5, _).

decideMenuAluno(6, _).

decideMenuAluno(7, Aluno):- 
    trocarSenha(Aluno),
    exibeMenuAluno(Aluno.id).

decideMenuAluno(8, _) :- write('\nDeslogando...'), menuPrincipal().

decideMenuAluno(_, _) :- msgInputInvalido().